plugins/utils/which-key.nix

```diff
@@ -55,257 +126,265 @@ with lib;
+
+        Note: the window options have changed parameters in v3 of which-key.
+      '';
+      visible = false;
+    };
+  };
+
+  extraConfig =
+    cfg:
     let
       cfg = config.plugins.which-key;
       opt = options.plugins.which-key;
-      setupOptions = {
-        plugins = {
-          inherit (cfg.plugins) marks registers spelling;
-          presets = {
-            inherit (cfg.plugins.presets) operators motions;
-            text_objects = cfg.plugins.presets.textObjects;
-            inherit (cfg.plugins.presets)
-              windows
-              nav
-              z
-              g
-              ;
-          };
-        };
-        inherit (cfg) operators;
-        key_labels = cfg.keyLabels;
-        inherit (cfg) motions icons;
-        popup_mappings = {
-          scroll_down = cfg.popupMappings.scrollDown;
-          scroll_up = cfg.popupMappings.scrollUp;
-        };
-        window =
-          let
-            mkSpacing =
-              opt:
-              helpers.ifNonNull' opt [
-                opt.top
-                opt.right
-                opt.bottom
-                opt.top
-              ];
-          in
-          {
-            inherit (cfg.window) border position;
-            margin = mkSpacing cfg.window.margin;
-            padding = mkSpacing cfg.window.padding;
-            inherit (cfg.window) winblend;
-          };
-        inherit (cfg) layout;
-        ignore_missing = cfg.ignoreMissing;
-        inherit (cfg) hidden;
-        show_help = cfg.showHelp;
-        show_keys = cfg.showKeys;
-        inherit (cfg.settings) spec;
-        inherit (cfg) triggers;
-        triggers_nowait = cfg.triggersNoWait;
-        triggers_blacklist = cfg.triggersBlackList;
-        inherit (cfg) disable;
-      };
     in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
+    {
+      warnings =
+        optional opt.registrations.isDefined ''
+          nixvim (plugins.which-key):
+          The option definition `plugins.which-key.registrations' in ${showFiles opt.registrations.files} has been deprecated in which-key v3; please remove it.
+          You should use `plugins.which-key.settings.spec' instead.
 
-      warnings = optional opt.registrations.isDefined ''
-        nixvim (plugins.which-key):
-        The option definition `plugins.which-key.registrations' in ${showFiles opt.registrations.files} has been deprecated in which-key v3; please remove it.
-        You should use `plugins.which-key.settings.spec' instead.
+          Note: the spec format has changed in which-key v3
+          See: https://github.com/folke/which-key.nvim?tab=readme-ov-file#%EF%B8%8F-mappings
+        ''
+        ++ optional opt.window.isDefined ''
+          nixvim (plugins.which-key):
+          The option definition `plugins.which-key.window' in ${showFiles opt.window.files} has been deprecated in which-key v3; please remove it.
+          You should use `plugins.which-key.settings.win' instead.
```

Comment:
This warning is great, but I think we still need to implement passing `cfg.window` through to which-key?

If it maps to a setting option, we can do something like:
```
plugins.which-key.settings = optionalAttrs opt.window.isDefined {
  win = lib.modules.mkAliasAndWrapDefsWithPriority lib.id opt.window;
};
```

`lib.id` (aka `x: x`) could be replaced with a transformation function if needed.

Reply:
Was thinking maybe like this:
```nix
    extraConfigLua =
      let
        setupOptions = optionalAttrs opt.window.isDefined {
          win = lib.modules.mkAliasAndWrapDefsWithPriority lib.id opt.window;
        };
      in
      ''
        require("which-key").setup({ ${toLuaObject setupOptions}})
      ''
      + optionalString (opt.registrations.isDefined && cfg.registrations != { }) ''
        require("which-key").register(${toLuaObject cfg.registrations})
      '';
```
But, this is what gets generated with that inside the module... which doesn't look right to me.

```lua
require("which-key").setup({
    key_labels = {
        ["<BS>"] = "BACKSPACE",
        ["<CR>"] = "RETURN",
        ["<TAB>"] = "TAB",
        ["<bs>"] = "BACKSPACE",
        ["<cr>"] = "RETURN",
        ["<leader>"] = "SPACE",
        ["<space>"] = "SPACE",
        ["<tab>"] = "TAB",
    },
    spec = {
        { "<leader>c", group = "󰄄 Codesnap", mode = "v" },
        { "<leader>d", desc = "  Debug", mode = "n" },
        { "<leader>gW", group = "󰙅 Worktree" },
        { "<leader>gh", group = "  Hunks" },
        { "<leader>ug", group = "󰊢 Git" },
        { "<leader>p", group = " Preview", mode = "n" },
        { "<leader>h", group = "󱡀 Harpoon" },
        { "<leader>ha", desc = "Add" },
        { "<leader>he", desc = "QuickMenu" },
        { "<leader>hj", desc = "1" },
        { "<leader>hk", desc = "2" },
        { "<leader>hl", desc = "3" },
        { "<leader>hm", desc = "4" },
        { "<leader>l", group = "  LSP" },
        { "<leader>la", desc = "Code Action" },
        { "<leader>ld", desc = "Definition" },
        { "<leader>lD", desc = "References" },
        { "<leader>lf", desc = "Format" },
        { "<leader>lp", desc = "Prev" },
        { "<leader>ln", desc = "Next" },
        { "<leader>lt", desc = "Type Definition" },
        { "<leader>li", desc = "Implementation" },
        { "<leader>lh", desc = "Hover" },
        { "<leader>lr", desc = "Rename" },
        { "<leader>r", group = " Refactor", mode = "x" },
        { "<leader>x", group = " Trouble", mode = "n" },
        { "<leader>b", group = "󰓩 Buffers" },
        { "<leader>bs", group = "󰒺 Sort" },
        { "<leader>g", group = "󰊢 Git" },
        { "<leader>f", group = " Find" },
        { "<leader>r", group = " Refactor" },
        { "<leader>t", group = " Terminal" },
        { "<leader>u", group = " UI/UX" },
    },
})

require("which-key").setup({
    {
        win = {
            _type = "if",
            condition = true,
            content = {
                _type = "merge",
                contents = { { _type = "override", content = { border = "single" }, priority = 100 } },
            },
        },
    },
})
```

Reply:
You're right it shouldn't be set as `settings.win`, since the old format had a different name.

We could take advantage of `settings` being freeform here though and do something like this
https://github.com/nix-community/nixvim/blob/2089eb407d8c5dbd6ca6e93d4988a439ca6446fd/plugins/utils/which-key.nix#L262-L278

But wrapped in an `optionalAttrs opt.window.isDefined {}`:

```nix
config = {
  settings = optionalAttrs opt.window.isDefined {
    window = makeAliasAndWrapWithPriority (
      V:
      let
        mkSpacing =
          x:
          helpers.ifNonNull' x [
            x.top
            x.right
            x.bottom
            x.top
          ];
      in
      {
        inherit (v) border position;
        margin = mkSpacing v.margin;
        padding = mkSpacing v.padding;
        inherit (v) winblend;
      }
    ) opt.window;
  };
}
```

(aliasing and wrapping with priority here is probably overkill, since this a deprecated option and is likely only set in the old location. I included it just for the example)

Reply:
The simple version of that example is:

```nix
config = {
  settings = optionalAttrs opt.window.isDefined {
    window =
    let
      mkSpacing =
        x:
        helpers.ifNonNull' x [
          x.top
          x.right
          x.bottom
          x.top
        ];
    in
    {
      inherit (cfg.window) border position;
      margin = mkSpacing cfg.window.margin;
      padding = mkSpacing cfg.window.padding;
      inherit (cfg.window) winblend;
    };
  };
}
```